### ISSUE TITLE
[MNG-6604] Intermittent failures while downloading GAVs from Nexus

### ISSUE THREAD

**Issue by jira-importer** [apache/maven#8607]
**[Ivan Rizzante](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=irizzant)** opened **[MNG-6604](https://issues.apache.org/jira/browse/MNG-6604?redirect=false)** and commented

Hello

we're running maven 3.6.0 builds in a docker container and we use Nexus OSS configured as proxy for Maven Central.

While running our builds using Gitlab CI, we're experiencing intermittent build failures because Maven cannot find artifacts in Nexus which we verified they are actually available.

Error example below:

```
20744 [main] [ERROR] Failed to execute goal on project EcotransitWSClient: Could not resolve dependencies for project it.sdb.ecotransit:EcotransitWSClient:jar:7.0.1-SNAPSHOT: Could not transfer artifact commons-beanutils:commons-beanutils:jar:1.9.3 from/to nexus (http://maven-repo.sdb.it:8081/repository/maven-public/): /builds/sdb/webportal/.m2/repository/commons-beanutils/commons-beanutils/1.9.3/commons-beanutils-1.9.3.jar.part (No such file or directory) -> [Help 1]
```

I attached the full maven build log and our Docker env settings.

We tried disabling the keep alive and also disabling the connection pooling but nothing seems to fix the issue.

 

 

---

**Affects:** 3.6.0

**Attachments:**
- [docker-env.txt](https://issues.apache.org/jira/secure/attachment/12960212/docker-env.txt) (_2.76 kB_)
- [log.txt](https://issues.apache.org/jira/secure/attachment/12960213/log.txt) (_5.03 MB_)

**Issue Links:**
- [MNG-2802](https://issues.apache.org/jira/browse/MNG-2802) Concurrent-safe access to local Maven repository
 (_**"is a child of"**_)
- [MRESOLVER-65](https://issues.apache.org/jira/browse/MRESOLVER-65) When the build is run in parallel there are unsafe concurrent changes in the local repo
 (_**"is caused by"**_)
- [MNG-4706](https://issues.apache.org/jira/browse/MNG-4706) Multithreaded building can create bad files for downloaded artifacts in local repository

- [MNG-6996](https://issues.apache.org/jira/browse/MNG-6996) Upgrade Maven Resolver to 1.6.1
 (_**"requires"**_)

**Backported to:** [4.0.0-alpha-2](https://github.com/apache/maven/milestone/113?closed=1), [3.8.2](https://github.com/apache/maven/milestone/103?closed=1)

5 votes, 11 watchers

**Comment by jira-importer** [comment]
**[Ivan Rizzante](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=irizzant)** commented

I suppose that could be a possible explanation.
I'll try removing -T1C option and see if that fixes the issue.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Yes, please try!

**Comment by jira-importer** [comment]
**[Ivan Rizzante](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=irizzant)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I followed your suggestion:

```
Please also try Takari's concurrent repo implementation. This might fix the issue for you.
```

it looks like it solved the issue for the moment, but I'd like to test it some more.

**Comment by jira-importer** [comment]
**[Ivan Rizzante](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=irizzant)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) I can confirm the issue is solved, for a whole day of build we've never experienced a single failure

Using Takari's concurrent repo implementation did the trick for me, thank you.

Is there any plan to integrate that repo implementation and/or fixing the existing one?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I cannot provide an ETA, but  several things are in the pipeline. I will leave this ticket open until we have solved it.

**Comment by jira-importer** [comment]
**[Steve Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stodorov)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) we are currently hitting the same issue and have a similar setup. Our builds are using `-T2C` and we've recently started having failing builds which fail at random artifacts with the following message (same as the OP's log):
.m2/repository/net/minidev/json-smart/2.3/json-smart-2.3.jar.part (No such file or directory) 
 

We are currently using Maven 3.5.4 and have tried to also use the latest 3.6.1 from the staging repository with no luck. Unfortunately the [takari-local-repository](https://github.com/takari/takari-local-repository) does not solve the issue for us - it only partially minimizes it. The only way this issue is "resolved" is by not running the build with `-T2C` at all. However that increases our build time roughly 2 times.

Are there any updates on this or plans to fix this rather critical issue?

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o),

We're really not happy about having to drop the `-T2C` option, as it will significantly increase our build times.

Would it be possible to include a fix for this in the upcoming `3.6.1`, or could you please postpone it until this has been resolved? In essence, this means that the parallel builds don't actually really work. Well, at least not when you need to use them in a CI server.

What further details would you need in order to investigate and fix this?

**Comment by jira-importer** [comment]
**[Oleksandr Gryniuk](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=oleksandrg)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)

I would also appreciate if you could include the fix for this bug into the next release.

Thanks for your attention

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

[jvanzyl](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=jvanzyl), [Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf),

As authors of the `takari-local-repository`, would you mind sharing your thoughts, or giving some hints on how we could overcome this? We have 24-core machines and we would like to use them to their maximum potential. We have a lot of parallelized tests as well using JUnit 5.

**Comment by jira-importer** [comment]
**[Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf)** commented

I have not looked at Maven for several years now, so can't comment on the current state of affairs. Here is what I remember from the distant past

talari-local-repo is broken, don't use use it. https://github.com/takari/takari-local-repository/issues/5 https://github.com/takari/takari-local-repository/issues/4. 

-Daether.connector.resumeDownload=false may help workaround aether local repo concurrency bugs.

You probably want to use https://github.com/takari/takari-smart-builder to build on multiple threads. https://github.com/takari/concurrent-build-logger maybe useful too.

Good luck.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[OleksandrG](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=OleksandrG), how am I supposed to add a fix for the next release if I don't have any?

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

[Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf),

Thanks for confirming that the `takari-local-repository` is broken (as we thought) and that it's not us misconfiguring something! :)
We tried using `-Daether.connector.resumeDownload=false`, but this, unfortunately, did not help.

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o),

What sort of information would help you guys further investigate this?

Do you guys have anything else you could advise that we try?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring), my problem isn't information, but time which I sparely have and this requires a significant amount of time to fix I absolutely cannot allocate.

The property name is `aether.connector.resumeDownloads` or per-repo `aether.connector.resumeDownloads.<repoId>`. See `BasicRepositoryConnector`.

**Comment by jira-importer** [comment]
**[Dan Tran](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=dantran)** commented

This is very strange,  I am very sure there are a lot of us successfully building maven projects with hundreds of modules using smart builder. So far I have not seen this issue even with the broken takari local-repo.  I am using artifactory thou

Could this be one of those build ordering issue at user side? This is  fixable via maven dependency

**Comment by jira-importer** [comment]
**[Steve Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=stodorov)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) many thanks for your previous comment! I didn't spot the typo - setting `-Daether.connector.resumeDownloads=true` seems to have solved the issue.

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

Many thanks for your help, [Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) and [Igor Fedorenko](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=igorf)!

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Glad, the workaround works for you. This issue isn't resolved, I will leave it open until someone can fix this.

**Comment by jira-importer** [comment]
**[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt)** commented

We are getting this issue as well however we are not using a Nexus repository but just a plain web server (nginx).

I presume it has something to do with the fact we are using the multithreaded build.

EDIT it appears that MNG-4706 is more likely the issue and thus enabling partial downloads will not help and perhaps exacerbate the issue.

Thus we might try

```
-Daether.connector.basic.threads=1
-Daether.connector.resumeDownloads=false
```

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt), I am quite certain that this is the case. Have you tried to reduce the amount of workers?

**Comment by jira-importer** [comment]
**[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) see my edits on previous comment.

We are using 1C on a 8 core machine (I believe.. it might be hyperthreaded so I don't know the actually count).

Our build machine is super duper fast so concurrency issues like this happen frequently. I really don't want to reduce the worker count as we like builds as fast as possible.

**Comment by jira-importer** [comment]
**[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt)** commented

Since this is happening on our build machine and our build machine has file system access to the repository I'm wondering what would happen if I just use file system access instead of HTTP wagon. I would imagine it would reduce the race condition but not eliminate it.

EDIT I also like to add that our build machine (jenkins) uses a separate repository per build so this is fairly annoying given all the separation we are already doing.

I love Maven but this critical problems that don't ever seem to fixed are making me reconsider Gradle.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I'll get back to you this evening and share some information where this issue is actually rooted.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt), I looked again into the code and exchanging the transport won't solve your problem. As far as I can see the problem is rooted in `org.eclipse.aether.connector.basic.PartialFile` in module `maven-resolver-connector-basic`.

We have actually two issues, but we must solve only one at the moment (#2).

Issue #1: Multiple Maven instances (processes) trying to write to the same local repo and the same file
Issue #2: Multiple threads within one Maven instance (process) trying to write to the same local repo and the same file

If someone wants really fix this, he needs a clean room impl of a proper concurrent file locking (akin to `https://commons.apache.org/proper/commons-io/javadocs/api-release/org/apache/commons/io/output/LockableFileWriter.html` outside of Maven Resolver, e.g., a simple `ExecutorService` which downloads the same file in parallel threads and they wait for each other. As soon as this locking works, it can be ported to Maven Resolver. My multithreading skills are low these days and I assumed that even a skilled dev will require at least 3 to 5 days to have a working solution for this problem. **Who can spare a week and brain power?**

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

Could this perhaps be solved by using one of these two options:
* Have a daemon for Maven that controls downloads. When an artifact needs to be resolved from a different instance of Maven running in a different JVM, it would connect to the daemon and see, if there already is a download for this in progress and wait for it.
* Have a temporary directory to which the artifacts are written, instead of local repository. When an artifact needs to be resolved from a different instance of Maven running in a different JVM, check the temporary directory for the artifact and check, if the size is growing every X milliseconds and wait. If there has been 15 seconds of inactivity. then the download is assumed dead and the querying instance of Maven, can feel free to restart the download.

`LockableFileWriter` is possibly another alternative.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The first point will only solve a multi VM solution, i.e, multiple instances will try to download the same file. If you have one Maven instance requesting the same artifact that daemon will use threads to download and you will simply push the issue to the deamon.
The second point can be implemented without a temp dir, but with proper locking. The mentioned `PartialFile` works with a wait mechanism which is obviously not so good.

A reliable detection mechanism has to do two things:

* Detect reliably that this file is locked
* Wait for a certain amount of time. This is hard because it could be a gigabyte large file and obviously you might need to wait minutes to download this one.

If we say that we generally do not support multi VM access to one local repo, we'd need centralized, in-memory keyed locks to synchronize on. We could even avoid repeated downloads.

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

In this day and age saying the such a widely used tool for building code like Maven will not support multiple VM access will be madness. I'm not a core Maven developer, but I've been using it and developing plugins and tooling around it since it first popped up many years ago. CI build farms need this to be properly implemented.

I think that we need to focus on implementing a proper locking mechanism that works for multiple JVM-s.

I get your point about the complexity of fine-tuning this. I don't think one should worry about large files, as much as tracking inactivity on files (a JVM may have terminated and left a file behind and then lock things for other JVM-s; if this has happened an easy way to check it would be to simply see when the last file write activity happened and see, if the `maxInactivity` time has been exceeded).

What are the classes of importance, if somebody would like to dig into this?

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

> I think that we need to focus on implementing a proper locking mechanism that works for multiple JVM-s.

I agree.

> What are the classes of importance, if somebody would like to dig into this?

I mentioned them a nfew comments earlier.

**Comment by jira-importer** [comment]
**[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt)** commented

[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o) My late thoughts (was on vacation):

You are correct in that most of the problems are in the [BasicRepositoryConnector](https://github.com/apache/maven-resolver/blob/maven-resolver-1.4.0/maven-resolver-connector-basic/src/main/java/org/eclipse/aether/connector/basic/BasicRepositoryConnector.java) and its dependencies (partial file). I accidentally said Wagon earlier but I meant the BasicRepositoryConnector.

Incidentally I do have quite a bit of experience in dealing with concurrency issues but I'm very reluctant to spend the time as I have had problems with uptake (ie accepting pull requests) with Apache projects even when following all the guidelines.

Furthermore if you look at MNG-2802 there have been attempts by others to try to fix the issue (or similar issues) but the turn around time and [communication seems to be lacking by those "in charge"](https://github.com/apache/maven-wagon/pull/49#issuecomment-431325340).

For now I will keep using:

```
-Daether.connector.basic.threads=1
```

and separate local repositories per build process (jenkins has configuration to do this) to I hope avoid these issues.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

[Adam Gent](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=agentgt), I understand myself that it is frustrating that no one looks at your PRs. I have the same with other projects. I try to review and merge as good as I can, but if one PR requires hours of work, this just will ned time. At the end this is volunteer work. Which PRs do you have open with Maven and friends I can look at?

I am aware of the PR for MNG-2802, but there seems to be unresolved issues with the PR. I can rereview it, but am not really fond of the inline `System#exit()`.

As sad as it sounds, please still to the properly for now.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

I have left an extended review on the PR, please add to the discussion if you like.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Please test a new, global locking approach with MRESOLVER-123. It is suited for concurrent access in **one** VM.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

MRESOLVER-131 introduces a fully new, multi-JVM concurrent approach based on Redis. Try it from here: https://github.com/apache/maven-resolver/tree/MRESOLVER-131. Read it from here: https://maven.apache.org/resolver-archives/resolver-LATEST/maven-resolver-synccontext-redisson/index.html

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Use any of the two new sync contexts to solve this concurrency issue.

**Comment by jira-importer** [comment]
**[Martin Todorov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=carlspring)** commented

The second link to MRESOLVER-131 is broken.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

The branch does not exist anymore and has been merged into master which in turn has been tagged as 1.6.1 and is on Central.

**Comment by jira-importer** [comment]
**[Michael Osipov](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=michael-o)** commented

Maven Resolver 1.6.2 will be available with Maven 3.8.2 so this has been addressed in 3.8.x too.
